ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Forge.hs

```diff
@@ -61,7 +73,50 @@ forgeShelleyBlock hotKey canBeLeader cfg curNo curSlot tickedLedger txs isLeader
     TPraosConfig { tpraosParams = TPraosParams { tpraosSlotsPerKESPeriod } } =
       configConsensus cfg
 
-    body = SL.toTxSeq @era $ Seq.fromList $ (\(ShelleyValidatedTx _ tx) -> tx) <$> txs
+    computedMaxTxCapacity = computeMaxTxCapacity tickedLedger maxTxCapacityOverride
+    computedMaxTxExUnitCapacity = maxTxExUnitCapacity tickedLedger
+
+    body =
+        SL.toTxSeq @era
+      . Seq.fromList
+      . toTxInBlock
+      $ toPayloads txs
+
+    toPayloads :: [Validated (GenTx (ShelleyBlock era))] -> BlockPayloads era
+    toPayloads = foldr (appendTx computedMaxTxCapacity) initBlockPayloads
+
+    initBlockPayloads :: BlockPayloads era
+    initBlockPayloads = BlockPayloads {
+        bpTxs  = []
+      , bpSize = 0
+      , bpExUnits = mempty
+    }
+
+    appendTx ::
+         Word32
+      -> (Validated (GenTx (ShelleyBlock era)))
+      -> BlockPayloads era
+      -> BlockPayloads era
+    appendTx maxBlockSize tx bp@(BlockPayloads bpTxs bpSize bpExUnits)
+      -- do not include transaction if it will exceed maximum block size
+      | bpSize + txInBlockSize genTx > maxBlockSize        = bp
+      -- do not include transaction if it will exceed maximum ex unit size
+      | exUnitsExceedsLimit (bpExUnits <> txExUnitsSize genTx) = bp
+      | otherwise = bp { bpTxs = tx : bpTxs
+                       , bpSize = txInBlockSize (txForgetValidated tx) + bpSize
+                       , bpExUnits = bpExUnits <> txExUnitsSize genTx
+                       }
+      where
```

Comment:
Similar to my comment on the Byron function, my instinct is to define `bp'` in the `where` clauses and use it in the guard conditions and as the guarded return value.